usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -5421,149 +5413,122 @@ static CK_RV dh_generate_keypair(STDLL_TokData_t * tokdata,
+                              dh_ulPublicKeyAttributeCount,
+                              &new_publ_attrs, &new_publ_attrs_len, -1);
     if (rc != CKR_OK) {
-        TRACE_ERROR("%s build_attribute failed with rc=0x%lx\n", __func__, rc);
+        TRACE_ERROR("%s DH check public key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
         goto dh_generate_keypair_end;
     }
-    rc = template_update_attribute(priv_tmpl, attr);
+
+    rc = check_key_attributes(tokdata, CKK_DH, CKO_PRIVATE_KEY,
+                              dh_pPrivateKeyTemplate,
+                              dh_ulPrivateKeyAttributeCount,
+                              &new_priv_attrs, &new_priv_attrs_len, -1);
     if (rc != CKR_OK) {
-        TRACE_ERROR("%s template_update_attribute failed with rc=0x%lx\n",
-                    __func__, rc);
-        free(attr);
+        TRACE_ERROR("%s DH check private key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
         goto dh_generate_keypair_end;
     }
-    rc = build_attribute(CKA_BASE, base_attr->pValue,
-                         base_attr->ulValueLen, &attr);
+
+    /* card does not want CKA_PRIME/CKA_BASE in template but in dh_pgs */
+    rc = template_attribute_get_non_empty(publ_tmpl, CKA_PRIME,
+                                          &prime_attr);
     if (rc != CKR_OK) {
-        TRACE_ERROR("%s build_attribute failed with rc=0x%lx\n", __func__, rc);
+        TRACE_ERROR("%s DH No CKA_PRIME attribute found\n", __func__);
         goto dh_generate_keypair_end;
     }
-    rc = template_update_attribute(priv_tmpl, attr);
+
+    rc = template_attribute_get_non_empty(publ_tmpl, CKA_BASE,
+                                          &base_attr);
     if (rc != CKR_OK) {
-        TRACE_ERROR("%s template_update_attribute failed with rc=0x%lx\n",
-                    __func__, rc);
-        free(attr);
+        TRACE_ERROR("%s DH No CKA_BASE attribute found\n", __func__);
         goto dh_generate_keypair_end;
     }
 
-    /* copy CKA_PRIME/CKA_BASE values */
-    dh_pgs.pg = malloc(p_len * 2);
+    dh_pgs.pg = malloc(prime_attr->ulValueLen * 2);
     if (!dh_pgs.pg) {
         TRACE_ERROR("%s Memory allocation failed\n", __func__);
         rc = CKR_HOST_MEMORY;
         goto dh_generate_keypair_end;
     }
-    memset(dh_pgs.pg, 0, p_len * 2);
-    memcpy(dh_pgs.pg, prime_attr->pValue, p_len);     /* copy CKA_PRIME value */
+
+    memset(dh_pgs.pg, 0, prime_attr->ulValueLen * 2);
+    /* copy CKA_PRIME value */
+    memcpy(dh_pgs.pg, prime_attr->pValue, prime_attr->ulValueLen);
     /* copy CKA_BASE value, it must have leading zeros
      * if it is shorter than CKA_PRIME
      */
-    memcpy(dh_pgs.pg + p_len + (p_len - g_len), base_attr->pValue, g_len);
-    dh_pgs.pg_bytes = p_len * 2;
+    memcpy(dh_pgs.pg + prime_attr->ulValueLen +
+                        (prime_attr->ulValueLen - prime_attr->ulValueLen),
+           base_attr->pValue, prime_attr->ulValueLen);
```

Comment:
Is the calculation correct? 
dh_pgs.pg + prime_attr->ulValueLen + (prime_attr->ulValueLen - prime_attr->ulValueLen)
Shouldn't it be (prime_attr->ulValueLen - base_attr->ulValueLen)?
And line 5486, base_attr->ulValueLen  instead of prime_attr->ulValueLen?

Reply:
Good catch. I seem to have messed this up when reworking the code.
Will fix.